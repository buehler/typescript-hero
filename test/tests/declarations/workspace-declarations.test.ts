import { parse, sep } from 'path';
import * as sinon from 'sinon';
import { DefaultDeclaration, File } from 'typescript-parser';
import { workspace } from 'vscode';

import WorkspaceDeclarations, { WorkspaceDeclarationsState } from '../../../src/declarations/workspace-declarations';
import { expect, wait } from '../setup';

describe('WorkspaceDeclarations', () => {

  const folder = workspace.workspaceFolders![0];
  let workspaceDeclarations: WorkspaceDeclarations;

  beforeEach(() => {
    workspaceDeclarations = new WorkspaceDeclarations(folder);
  });

  afterEach(() => {
    if (workspaceDeclarations) {
      workspaceDeclarations.dispose();
    }
  });

  it('should create an object', () => {
    expect(new WorkspaceDeclarations(folder)).to.exist;
  });

  it('should return the correct files', async () => {
    const files = await (workspaceDeclarations as any).findFiles() as string[];
    const declarationFiles = files
      .filter(file => file.indexOf(`workspace_1${sep}declarations`) >= 0)
      .map(file => parse(file).base);

    expect(declarationFiles).to.matchSnapshot();
  });

  it('should not add non ts files', async () => {
    const files = await (workspaceDeclarations as any).findFiles() as string[];
    const declarationFiles = files
      .filter(file => file.indexOf(`workspace_1${sep}declarations`) >= 0)
      .map(file => parse(file).ext);

    for (const ext of declarationFiles) {
      expect(ext).to.equal('.ts');
    }
  });

  it('should initialize an index', async () => {
    expect(workspaceDeclarations.index).not.to.exist;
    await workspaceDeclarations.initialize();
    expect(workspaceDeclarations.index).to.exist;
    expect(workspaceDeclarations.index.indexReady).to.equal(true);
  });

  it('should call syncing event', async () => {
    const spy = sinon.spy();
    workspaceDeclarations.workspaceStateChanged(spy);
    await workspaceDeclarations.initialize();
    expect(spy.getCall(0).args[0]).to.equal(WorkspaceDeclarationsState.Syncing);
  });

  it('should call idle event on finished sync', async () => {
    const spy = sinon.spy();
    workspaceDeclarations.workspaceStateChanged(spy);
    await workspaceDeclarations.initialize();
    expect(spy.getCall(1).args[0]).to.equal(WorkspaceDeclarationsState.Idle);
  });

  it('should call reindex when a file has changed', async () => {
    await workspaceDeclarations.initialize();
    const eventSpy = sinon.spy();
    const reindexSpy = sinon.stub(workspaceDeclarations.index, 'reindexForChanges');
    reindexSpy.returns(Promise.resolve());
    workspaceDeclarations.workspaceStateChanged(eventSpy);

    expect(eventSpy.callCount).to.equal(0);
    (workspaceDeclarations as any).fileWatcherEvent(
      'updated',
      { fsPath: './file-to-index.ts' },
    );
    await wait(550);
    expect(eventSpy.callCount).to.equal(2);
    expect(reindexSpy.getCall(0).args).to.matchSnapshot();
  });

  it('should return the correct declarations for a workspace', async () => {
    await workspaceDeclarations.initialize();

    const declarations = workspaceDeclarations.index.declarationInfos.filter(d => d.from.startsWith('/declarations'));
    for (const d of declarations) {
      if (d.declaration instanceof DefaultDeclaration) {
        delete ((d.declaration as any).resource as File).filePath;
        delete (d.declaration as any).resource.rootPath;
      }
    }
    expect(declarations).to.matchSnapshot();
  });

  it('should return the correct declarations for a file', async () => {
    await workspaceDeclarations.initialize();

    const declarations = workspaceDeclarations
      .index
      .declarationInfos
      .filter(d => d.from.startsWith('/declarations/multi-declarations'));

    expect(declarations).to.matchSnapshot();
  });

  it('should correctly move declarations for a barrel file', async () => {
    await workspaceDeclarations.initialize();

    const declarations1 = workspaceDeclarations
      .index
      .declarationInfos
      .filter(d => d.from.startsWith('/declarations/declarations-for-barrel-1'));
    const declarations2 = workspaceDeclarations
      .index
      .declarationInfos
      .filter(d => d.from.startsWith('/declarations/declarations-for-barrel-2'));
    const barrel = workspaceDeclarations
      .index
      .declarationInfos
      .filter(d => d.from.startsWith('/declarations/barrel'));

    expect(declarations1).to.matchSnapshot();
    expect(declarations2).to.matchSnapshot();
    expect(barrel).to.matchSnapshot();
  });

  it('should correctly remove trailing /index from location', async () => {
    await workspaceDeclarations.initialize();

    expect(
      workspaceDeclarations.index.declarationInfos
        .filter(d => d.from.startsWith('/declarations'))
        .map(d => d.from)
        .every(from => from.indexOf('/index') === -1),
    ).to.equal(true);
  });

  it('should correctly ignore workspace files that are ignored by config');

  it('should correctly add workspace files that are not ignored by config anymore');

});
