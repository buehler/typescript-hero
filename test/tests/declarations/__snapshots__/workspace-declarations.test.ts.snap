// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceDeclarations should call reindex when a file has changed 1`] = `
Array [
  Object {
    "created": Array [],
    "deleted": Array [],
    "updated": Array [
      "./file-to-index.ts",
    ],
  },
]
`;

exports[`WorkspaceDeclarations should correctly move declarations for a barrel file 1`] = `Array []`;

exports[`WorkspaceDeclarations should correctly move declarations for a barrel file 2`] = `Array []`;

exports[`WorkspaceDeclarations should correctly move declarations for a barrel file 3`] = `
Array [
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 32,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel1",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 65,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel2",
      "properties": Array [],
      "start": 33,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 98,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel3",
      "properties": Array [],
      "start": 66,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 131,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel4",
      "properties": Array [],
      "start": 99,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 164,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel5",
      "properties": Array [],
      "start": 132,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": EnumDeclaration {
      "end": 204,
      "isExported": true,
      "members": Array [],
      "name": "EnumForBarrel",
      "start": 175,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 205,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel1",
      "properties": Array [],
      "start": 165,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 246,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel2",
      "properties": Array [],
      "start": 206,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 287,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel3",
      "properties": Array [],
      "start": 247,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 34,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel1",
      "start": 0,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 69,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel2",
      "start": 35,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 104,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel3",
      "start": 70,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 139,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel4",
      "start": 105,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 174,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel5",
      "start": 140,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
]
`;

exports[`WorkspaceDeclarations should return the correct declarations for a file 1`] = `
Array [
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 224,
      "isExported": true,
      "methods": Array [
        MethodDeclaration {
          "end": 191,
          "isAbstract": false,
          "name": "aMethod",
          "parameters": Array [
            ParameterDeclaration {
              "end": 159,
              "name": "str",
              "start": 148,
              "type": "string",
            },
          ],
          "start": 130,
          "type": "T",
          "variables": Array [],
          "visibility": 2,
        },
        MethodDeclaration {
          "end": 222,
          "isAbstract": false,
          "name": "bMethod",
          "parameters": Array [],
          "start": 195,
          "type": "void",
          "variables": Array [],
          "visibility": 0,
        },
      ],
      "name": "Class",
      "properties": Array [
        PropertyDeclaration {
          "end": 126,
          "name": "prop",
          "start": 106,
          "type": "string",
          "visibility": 2,
        },
      ],
      "start": 83,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": EnumDeclaration {
      "end": 333,
      "isExported": true,
      "members": Array [
        "a",
        "b",
        "c",
      ],
      "name": "Enum",
      "start": 298,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 81,
      "isExported": true,
      "methods": Array [
        MethodDeclaration {
          "end": 47,
          "isAbstract": true,
          "name": "method1",
          "parameters": Array [],
          "start": 31,
          "type": "void",
          "variables": Array [],
          "visibility": 2,
        },
        MethodDeclaration {
          "end": 79,
          "isAbstract": true,
          "name": "method2",
          "parameters": Array [
            ParameterDeclaration {
              "end": 69,
              "name": "str",
              "start": 58,
              "type": "string",
            },
          ],
          "start": 50,
          "type": "number",
          "variables": Array [],
          "visibility": 2,
        },
      ],
      "name": "Interface",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 296,
      "isConst": true,
      "isExported": true,
      "name": "exportedConstant",
      "start": 261,
      "type": undefined,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 259,
      "isConst": false,
      "isExported": true,
      "name": "exportedVariable",
      "start": 226,
      "type": undefined,
    },
    "from": "/declarations/multi-declarations",
  },
]
`;

exports[`WorkspaceDeclarations should return the correct declarations for a workspace 1`] = `
Array [
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 224,
      "isExported": true,
      "methods": Array [
        MethodDeclaration {
          "end": 191,
          "isAbstract": false,
          "name": "aMethod",
          "parameters": Array [
            ParameterDeclaration {
              "end": 159,
              "name": "str",
              "start": 148,
              "type": "string",
            },
          ],
          "start": 130,
          "type": "T",
          "variables": Array [],
          "visibility": 2,
        },
        MethodDeclaration {
          "end": 222,
          "isAbstract": false,
          "name": "bMethod",
          "parameters": Array [],
          "start": 195,
          "type": "void",
          "variables": Array [],
          "visibility": 0,
        },
      ],
      "name": "Class",
      "properties": Array [
        PropertyDeclaration {
          "end": 126,
          "name": "prop",
          "start": 106,
          "type": "string",
          "visibility": 2,
        },
      ],
      "start": 83,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 32,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel1",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 65,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel2",
      "properties": Array [],
      "start": 33,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 98,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel3",
      "properties": Array [],
      "start": 66,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 131,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel4",
      "properties": Array [],
      "start": 99,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 164,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForBarrel5",
      "properties": Array [],
      "start": 132,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 31,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForIndex1",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 63,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForIndex2",
      "properties": Array [],
      "start": 32,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 95,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForIndex3",
      "properties": Array [],
      "start": 64,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 127,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForIndex4",
      "properties": Array [],
      "start": 96,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 159,
      "isExported": true,
      "methods": Array [],
      "name": "ClassForIndex5",
      "properties": Array [],
      "start": 128,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 72,
      "isExported": true,
      "methods": Array [],
      "name": "ClassInIndex",
      "properties": Array [],
      "start": 43,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": DefaultDeclaration {
      "end": undefined,
      "isExported": true,
      "name": "DefaultExportedClass",
      "resource": File {
        "declarations": Array [
          [Circular],
        ],
        "end": 46,
        "exports": Array [],
        "imports": Array [],
        "resources": Array [],
        "start": 0,
        "usages": Array [],
      },
      "start": undefined,
    },
    "from": "/declarations/default-export",
  },
  DeclarationInfo {
    "declaration": DefaultDeclaration {
      "end": undefined,
      "isExported": true,
      "name": "DefaultMultiExport",
      "resource": File {
        "declarations": Array [
          InterfaceDeclaration {
            "end": 50,
            "isExported": true,
            "methods": Array [],
            "name": "InterfaceInDefaultMultiExport",
            "properties": Array [],
            "start": 0,
          },
          [Circular],
        ],
        "end": 95,
        "exports": Array [],
        "imports": Array [],
        "resources": Array [],
        "start": 0,
        "usages": Array [],
      },
      "start": undefined,
    },
    "from": "/declarations/default-multi-export",
  },
  DeclarationInfo {
    "declaration": EnumDeclaration {
      "end": 333,
      "isExported": true,
      "members": Array [
        "a",
        "b",
        "c",
      ],
      "name": "Enum",
      "start": 298,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": EnumDeclaration {
      "end": 204,
      "isExported": true,
      "members": Array [],
      "name": "EnumForBarrel",
      "start": 175,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 81,
      "isExported": true,
      "methods": Array [
        MethodDeclaration {
          "end": 47,
          "isAbstract": true,
          "name": "method1",
          "parameters": Array [],
          "start": 31,
          "type": "void",
          "variables": Array [],
          "visibility": 2,
        },
        MethodDeclaration {
          "end": 79,
          "isAbstract": true,
          "name": "method2",
          "parameters": Array [
            ParameterDeclaration {
              "end": 69,
              "name": "str",
              "start": 58,
              "type": "string",
            },
          ],
          "start": 50,
          "type": "number",
          "variables": Array [],
          "visibility": 2,
        },
      ],
      "name": "Interface",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 205,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel1",
      "properties": Array [],
      "start": 165,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 246,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel2",
      "properties": Array [],
      "start": 206,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 287,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForBarrel3",
      "properties": Array [],
      "start": 247,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 199,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForIndex1",
      "properties": Array [],
      "start": 160,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 239,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForIndex2",
      "properties": Array [],
      "start": 200,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 279,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceForIndex3",
      "properties": Array [],
      "start": 240,
    },
    "from": "/declarations",
  },
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "end": 50,
      "isExported": true,
      "methods": Array [],
      "name": "InterfaceInDefaultMultiExport",
      "properties": Array [],
      "start": 0,
    },
    "from": "/declarations/default-multi-export",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 34,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel1",
      "start": 0,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 69,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel2",
      "start": 35,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 104,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel3",
      "start": 70,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 139,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel4",
      "start": 105,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 174,
      "isConst": true,
      "isExported": true,
      "name": "constForBarrel5",
      "start": 140,
      "type": undefined,
    },
    "from": "/declarations/barrel",
  },
  DeclarationInfo {
    "declaration": DefaultDeclaration {
      "end": undefined,
      "isExported": true,
      "name": "default-unnamed-export",
      "resource": File {
        "declarations": Array [
          [Circular],
        ],
        "end": 25,
        "exports": Array [],
        "imports": Array [],
        "resources": Array [],
        "start": 0,
        "usages": Array [],
      },
      "start": undefined,
    },
    "from": "/declarations/default-unnamed-export",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 296,
      "isConst": true,
      "isExported": true,
      "name": "exportedConstant",
      "start": 261,
      "type": undefined,
    },
    "from": "/declarations/multi-declarations",
  },
  DeclarationInfo {
    "declaration": VariableDeclaration {
      "end": 259,
      "isConst": false,
      "isExported": true,
      "name": "exportedVariable",
      "start": 226,
      "type": undefined,
    },
    "from": "/declarations/multi-declarations",
  },
]
`;

exports[`WorkspaceDeclarations should return the correct files 1`] = `
Array [
  "declarations-for-barrel-1.ts",
  "multi-declarations.ts",
  "barrel.ts",
  "default-export.ts",
  "declarations-for-index.ts",
  "default-multi-export.ts",
  "index.ts",
  "default-unnamed-export.ts",
  "declarations-for-barrel-2.ts",
]
`;
